pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -0,0 +1,751 @@
+		}
+	}
+}
+
+func (s *ReplicaSchedulingPreferenceController) deliverObj(obj pkgruntime.Object, delay time.Duration, failed bool) {
+	// TODO: right now this works on the RSP being same "ns/name" as the
+	// deploymenttemplate. Update this to rsp being able to use
+	// targetRef (kind+name) with multiple kinds.
+	// For reviewers - Plan is to use Objectmeta.OwnerReferences by setting it to
+	// created rsp into the target object. This way when reconcile is called
+	// the information will be available to link objects both ways.
+	qualifiedName := util.NewQualifiedName(obj)
+	s.deliver(qualifiedName, delay, failed)
+}
+
+// Adds backoff to delay if this delivery is related to some failure. Resets backoff if there was no failure.
+func (s *ReplicaSchedulingPreferenceController) deliver(qualifiedName util.QualifiedName, delay time.Duration, failed bool) {
+	key := qualifiedName.String()
+	if failed {
+		s.backoff.Next(key, time.Now())
+		delay = delay + s.backoff.Get(key)
+	} else {
+		s.backoff.Reset(key)
+	}
+	s.deliverer.DeliverAfter(key, &qualifiedName, delay)
+}
+
+// Check whether all data stores are in sync. False is returned if any of the informer/stores is not yet
+// synced with the corresponding api server.
+func (s *ReplicaSchedulingPreferenceController) isSynced() bool {
+	if !s.targetInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+
+	clusters, err := s.targetInformer.GetReadyClusters()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to get ready clusters: %v", err))
+		return false
+	}
+	if !s.targetInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	return true
+}
+
+// The function triggers reconciliation of all target federated resources.
+func (s *ReplicaSchedulingPreferenceController) reconcileOnClusterChange() {
+	if !s.isSynced() {
+		s.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(s.clusterAvailableDelay))
+	}
+	for _, obj := range s.templateStore.List() {
+		qualifiedName := util.NewQualifiedName(obj.(pkgruntime.Object))
+		s.deliver(qualifiedName, s.smallDelay, false)
+	}
+}
+
+func (s *ReplicaSchedulingPreferenceController) reconcile(qualifiedName util.QualifiedName) reconciliationStatus {
+	if !s.isSynced() {
+		return statusNotSynced
+	}
+
+	kind := "ReplicaSchedulingPreference"
+	key := qualifiedName.String()
+
+	glog.V(4).Infof("Starting to reconcile %v %v", kind, key)
+	startTime := time.Now()
+	defer glog.V(4).Infof("Finished reconciling %v %v (duration: %v)", kind, key, time.Now().Sub(startTime))
+
+	rsp, err := s.objFromCache(s.store, kind, key)
+	if err != nil {
+		return statusError
+	}
+	if rsp == nil {
+		return statusAllOK
+	}
+
+	typedRsp, ok := rsp.(*fedschedulingv1a1.ReplicaSchedulingPreference)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Incorrect runtime object for RSP: %v", rsp))
+		return statusNotSynced
+	}
+
+	if typedRsp.ObjectMeta.DeletionTimestamp != nil {
```

Comment:
Without a finalizer or the cleanup on deletion that would suggest one, consider removing this block and the `delete()` function.

Reply:
Updated!